Source/Scene/Cesium3DTilesetTraversal.js

```diff
@@ -0,0 +1,823 @@
+        this.tileset = tileset;
+        this.frameState = frameState;
+        this.outOfCore = outOfCore;
+        this.leaves.length = 0;
+        this.baseScreenSpaceError = Math.max(baseScreenSpaceError, this.tileset._maximumScreenSpaceError);
+        this.internalDFS.tileset = this.tileset;
+        this.internalDFS.frameState = this.frameState;
+        this.internalDFS.outOfCore = this.outOfCore;
+        this.internalDFS.baseScreenSpaceError = this.baseScreenSpaceError;
+        DFS(root, this);
+    };
+
+    BaseTraversal.prototype.visitStart = function(tile) {
+        if (tile._lastVisitedFrame !== this.frameState.frameNumber) {
+            visitTile(this.tileset, tile, this.frameState, this.outOfCore);
+        }
+    };
+
+    BaseTraversal.prototype.visitEnd = function(tile) {
+        tile._lastVisitedFrame = this.frameState.frameNumber;
+    };
+
+    BaseTraversal.prototype.getChildren = function(tile) {
+        if (baseUpdateAndCheckChildren(this.tileset, tile, this.baseScreenSpaceError, this.frameState)) {
+            var children = tile.children;
+            var childrenLength = children.length;
+            var allReady = true;
+            var replacementWithContent = tile.refine === Cesium3DTileRefine.REPLACE && tile.hasRenderableContent;
+            for (var i = 0; i < childrenLength; ++i) {
+                var child = children[i];
+                loadTile(child, this.frameState);
+                touch(this.tileset, child, this.outOfCore);
+
+                // content cannot be replaced until all of the nearest descendants with content are all loaded
+                if (replacementWithContent) {
+                    if (!child.hasEmptyContent) {
+                        allReady = allReady && child.contentReady;
+                    } else {
+                        allReady = allReady && this.internalDFS.execute(child);
+                    }
+                }
+            }
+
+            if (allReady) {
+                return children;
+            }
+        }
+        return emptyArray;
+    };
+
+    function baseUpdateAndCheckChildren(tileset, tile, baseScreenSpaceError, frameState) {
+        if (tile.hasTilesetContent) {
+            // load any tilesets of tilesets now because at this point we still have not achieved a base level of content
+            loadTile(tile, frameState);
+            if (tile.contentReady) {
+                updateChildren(tileset, tile, frameState);
+            }
+            return true;
+        }
+
+        if (hasAdditiveContent(tile)) {
+            tileset._desiredTiles.push(tile);
+        }
+
+        // stop traversal when we've attained the desired level of error
+        if (tile._screenSpaceError <= baseScreenSpaceError) {
+            // When skipping LODs, require an existing base level of content first
+            updateChildren(tileset, tile, frameState);
```

Comment:
`leafHandler` checks if at least one child is visible for the children union bound optimization